account_billing/models/account_billing.py

```diff
@@ -0,0 +1,257 @@
+                rec.name = (
+                    self.env["ir.sequence"]
+                    .with_context(ir_sequence_date=rec.date)
+                    .next_by_code(sequence_code)
+                )
+            rec.write({"state": "billed"})
+            rec.message_post(body=_("Billing is billed."))
+        return True
+
+    def action_cancel_draft(self):
+        for rec in self:
+            rec.write({"state": "draft"})
+            rec.message_post(body=_("Billing is reset to draft"))
+        return True
+
+    def action_cancel(self):
+        for rec in self:
+            invoice_paid = rec.billing_line_ids.mapped("move_id").filtered(
+                lambda m: m.payment_state == "paid"
+            )
+            if invoice_paid:
+                raise ValidationError(_("Invoice paid already."))
+            rec.write({"state": "cancel"})
+            self.message_post(body=_("Billing %s is cancelled") % rec.name)
+        return True
+
+    def action_register_payment(self):
+        return self.mapped("billing_line_ids.move_id").action_register_payment()
+
+    def invoice_relate_billing_tree_view(self):
+        name = self.bill_type == "out_invoice" and "Invoices" or "Bills"
+        return {
+            "name": _("%s") % (name),
+            "view_mode": "list,form",
+            "res_model": "account.move",
+            "view_id": False,
+            "views": [
+                (self.env.ref("account.view_move_tree").id, "list"),
+                (self.env.ref("account.view_move_form").id, "form"),
+            ],
+            "type": "ir.actions.act_window",
+            "domain": [("id", "in", [rec.move_id.id for rec in self.billing_line_ids])],
+            "context": {"create": False},
+        }
+
+    def _get_billing_line_dict(self, moves):
+        billing_line_dict = [
+            {
+                "billing_id": self.id,
+                "move_id": m.id,
+                "amount_total": m.amount_total
+                * (-1 if m.move_type in ["out_refund", "in_refund"] else 1),
+            }
+            for m in moves
+        ]
+        return billing_line_dict
+
+    def compute_lines(self):
+        self.billing_line_ids = False
+        types = ["in_invoice", "in_refund"]
+        if self.bill_type == "out_invoice":
+            types = ["out_invoice", "out_refund"]
+        moves = self._get_moves(self.threshold_date_type, types)
+        billing_line_dict = self._get_billing_line_dict(moves)
+        self.billing_line_ids.create(billing_line_dict)
+
+
+class AccountBillingLine(models.Model):
+    _name = "account.billing.line"
+    _description = "Billing Line"
+
+    billing_id = fields.Many2one(comodel_name="account.billing")
+    move_id = fields.Many2one(
+        comodel_name="account.move",
+        index=True,
+    )
+    name = fields.Char(related="move_id.name")
+    invoice_date = fields.Date(related="move_id.invoice_date")
+    threshold_date = fields.Date(related="move_id.invoice_date_due")
```

Comment:
I find the field name `threshold_date` quite misleading. I'd consolidate these two into a compute field (with `invoice_date` a surviving one), and simplify the relevant part of `validate_billing()`.

Reply:
Done!